SKIRT/core/SecondarySourceSystem.cpp

```diff
@@ -421,40 +421,76 @@ namespace
         {
             // first generate a random azimuth angle
             const double phi = 2. * M_PI * random->uniform();
-            // compute its sine and cosine
-            const double cosphi = cos(phi);
-            const double sinphi = sin(phi);
 
             // now generate a random zenith angle
             // first, compute the cumulative distribution for a photon at the given wavelength
-            // create an empty array with the same size as the zenith angle grid
+            // create empty PDF and CDF arrays with the same size as the zenith angle grid
             const Array& thetas = _ms->mix(_m, _hv[0])->thetaGrid();
-            Array cdf(thetas.size());
+            Array pdf(thetas.size()), cdf(thetas.size());
+            ;
             // add the distributions for the different components
             for (int h : _hv)
             {
                 const Array& Qabs = _ms->mix(_m, h)->sectionsAbs(_lambda);
-                cdf += _ms->numberDensity(_m, h) * Qabs;
+                pdf += _ms->numberDensity(_m, h) * Qabs;
```

Comment:
Just curious, why is there a cumulative += operator for calculating the pdf?

Reply:
Because the total PDF is the density-weighted average of the PDFs for different species in the mix.